Add event cancellation fields

shandy-bixly opened the pull request.

shandy-bixly commented on tendenci/apps/site_settings/management/commands/add_settings.json:
```diff
@@ -1919,5 +1919,22 @@
+        "update_dt": "2023-03-17 18:59:00",
+        "input_type": "select",
+        "client_editable": true,
+        "name": "allow_refunds",
```
sounds good!

shandy-bixly replied:
updated

jennyq replied:
Thanks!

jennyq reviewed (changes requested):
It seems there are some issues with the `prepare_value` method in apps/base/fields.py PercentField class.  On event add, it first throws a TypeError: unsupported operand type(s) for *: 'NoneType' and 'int'. And if non-numeric chars are entered in the "Cancellation Percent" field, it throws a ValueError: invalid literal for int() with base 10.

Looks like it is due to the `prepare_value` gets called before the validation, as a result, the check for `if self.error_occurred` always returns False ..

jennyq reviewed (changes requested):
Hi Shandy, 
More items need your attention. Thanks!
1) Please add refund info under the "Payment History" on invoice view.
2) Can you please change the cancellation fee in "A cancellation fee of $20.0000 will be applied to your invoice" to have 2 digits instead of 4 after the decimal point?
3) Accounting entries needs a little adjustment as well. Currently, the cancellation fee is not accounted.
4) Both anonymous users and logged_in users received a 404 page after using the cancel link with "registrants/cancel" in it (with auto refund on). Example url: http://127.0.0.1:8000/events/2/registrants/cancel/8221435bcce913b5c2dc22eaf6cb6590/. 
5) The cancel link with "registrations/cancel/" in it, for example, http://127.0.0.1:8000/events/2/registrations/cancel/56b7a720568f7e7e89904d13b53584b6/ worked, but as an anonymous user, I kept getting the alert message "Refund in the amount of $100.00 failed to process. Please contact support." after the cancellation, even though the invoice showed refund was a success. For the logged_in users, that error message didn't show.
6) Template code needs to be internationalized with translate and blocktranslate template tags. The texts in the code need to be marked as translation strings as well, for example,
```
    CHOICES = (
            (PENDING, _('Pending')),
            (SUCCEEDED, _('Succeeded')),
            (FAILED, _('Failed'))
     )
```

shandy-bixly commented:
> * e the cancellation fee in "A cancellation fee of $20.0000 will be applied to your invoice" to have 2 digits instead of 4 after the decimal point?
> * Accounting entries needs a little adjustment as well. Currently, the cancellation fee is not acco

I'm working through these, and on the accounting entries for cancellation fees I have a question:
    Since we're not actually creating a payment for cancellation fees when refunding (we just deduct the fees from the amount refunded) do we need to make accounting entries for a payment that's not actually ever made? If so, do we need to add the fees back into the accounting entries for the refunded amount? Ex. If the total charge is $100, with a $20 cancellation fee: 1) User pays registration of $100 (balance is $0, accounting entries are made for the $100 payment) 2) User cancels registration, and fee is added. (balance is $20) 3a) [auto refunds] $80 is refunded ($100- $20, accounting entries are made for the $80 refund, balance is $0) 3b) [manual refunds] Admin user refunds $80 (total refundable amount) (balance is $0, accounting entries are made for the $80 refund) 3c) [no refunds] User pays $20 (balance is $0, accounting entries made for the $20 payment)

shandy-bixly commented:
> * Both anonymous users and logged_in users received a 404 page after using the cancel link with "registrants/cancel" in it (with auto refund on). Example url: http://127.0.0.1:8000/events/2/registrants/cancel/8221435bcce913b5c2dc22eaf6cb6590/.
> * The cancel link with "registrations/cancel/" in it, for example, http://127.0.0.1:8000/events/2/registrations/cancel/56b7a720568f7e7e89904d13b53584b6/ worked, but as an anonymous user, I kept getting the alert message "Refund in the amount of $100.00 failed to process. Please contact support." after the cancellation, even though the invoice showed refund was a success. For the logged_in users, that error message didn't show.

I not able to repro these

jennyq commented:
RE accounting entries for refund, after some research and thinking, I added a new account "Sales Returns and Allowances" and the function "make_acct_entries_refund".  Also updated invoice.__refund to use it. https://github.com/tendenci/tendenci/pull/1159/commits/66245776be0b9e73cf393e8474e76ecabafdb691

jennyq commented:
> I not able to repro these

The 404 seems gone. Now I'm not able to reproduce it either. 

For the alert message "Refund in the amount of $100.00 failed to process. Please contact support.", I still experienced it when testing as an anonymous user. It's been fixed now.

jennyq commented:
To-Do List:

1) Need to make it work with stripe connect for both standard and express accounts.
2) "Refund Confirmation Failure" email (to admin) needs to include the reason why failed. Otherwise it would be troublesome to troubleshoot. Currently it has only one sentence "Refund to (name) in amount $100.00 failed to process through Stripe for (event name) on 06/23/2023." which provides no clue why it failed.
3) If Admin process a refund by clicking the "Refund" button on the invoice view, should it have the option to cancel the registration in the meantime? Currently it does not cancel the associated event registration. 
4) Need to update invoice report /invoices/report/overview/ to accommodate the added Refund table (to subtract the amount refunded).

shandy-bixly commented:
> To-Do List:
> 
> 1. Need to make refund work with stripe connect for both standard and express accounts.
> 2. "Refund Confirmation Failure" email (to admin) needs to include the reason why failed. Otherwise it would be troublesome to troubleshoot. Currently it has only one sentence "Refund to (name) in amount $100.00 failed to process through Stripe for (event name) on 06/23/2023." which provides no clue why it failed.
> 3. If Admin process a refund by clicking the "Refund" button on the invoice view, should it have the option to cancel the registration in the meantime? Currently it does not cancel the associated event registration.
> 4. Need to update invoice reports /invoices/report/overview/ and /reports/add/ to accommodate the added Refund table (to subtract the amount refunded).

Still working on point #4 (should have that done soon though)... pushing up what I have so far. I'm not sure what to look for to make sure Stripe Connect is working (other than there being no error)... with a Standard account, the refund processed without error. However, with Express, it told me the account didn't exist. This could be because I had just created the account... I will try again later to see if accepts it.

jennyq commented:
>  I'm not sure what to look for to make sure Stripe Connect is working (other than there being no error)... with a Standard account, the refund processed without error. However, with Express, it told me the account didn't exist. This could be because I had just created the account... I will try again later to see if accepts it.

I've adjusted params for refunding on the stripe express connected accounts https://github.com/tendenci/tendenci/pull/1159/commits/632aa0b22084472b43b5002e4b60e7112830c99a. It should be working now.

jennyq commented:
The refund via the Refund button on invoice view doesn't seem to be working as expected. 

Case 1:
As an admin, I went in to refund and cancel a registration (in the meantime) via the Refund button on the invoice view. As a result, two Refund rows were created under invoice description and two Refunded under Payment History.
![image](https://github.com/tendenci/tendenci/assets/106448/7d117c20-c139-4a30-b7ae-5c1c964d2294)


Case 2:
A registration was canceled, but not refunded due to an error. I went in as an admin to refund via the Refund button on the invoice view. Since the cancellation fee $2 had already been calculated and listed on the invoice, I specified the refund amount $18. But the actual amount refunded was $16. 
![image](https://github.com/tendenci/tendenci/assets/106448/8251b7e0-9195-4aa7-8fda-7162f9aba6ee)

shandy-bixly commented:
> The refund via the Refund button on invoice view doesn't seem to be working as expected.
> 
> Case 1: As an admin, I went in to refund and cancel a registration (in the meantime) via the Refund button on the invoice view. As a result, two Refund rows were created under invoice description and two Refunded under Payment History. ![image](https://user-images.githubusercontent.com/106448/246698510-7d117c20-c139-4a30-b7ae-5c1c964d2294.png)
> 
> Case 2: A registration was canceled, but not refunded due to an error. I went in as an admin to refund via the Refund button on the invoice view. Since the cancellation fee $2 had already been calculated and listed on the invoice, I specified the refund amount $18. But the actual amount refunded was $16. ![image](https://user-images.githubusercontent.com/106448/246698302-8251b7e0-9195-4aa7-8fda-7162f9aba6ee.png)

I suspect Case 1 came from code added to allow cancellation from the Refund menu. I will take a look. Probably somethiing small that can be fixed.

On Case 2, the cancelleation fee is automatically deducted provided you are refunding an amount larger than the fee. So if you specific $20. it will refund 18 (thus handling the $2 charge automatically by not refunding the $2).  Do we want the admin to decide whether or not they refund the cancellation fee? If so, then there would be a pending balance that the customer would need to pay instead of the fee being deducted from the refund.

shandy-bixly commented:
"Ideally, when a user cancels a registration, the total ($20) should remain unchanged instead of having the cancellation fee ($2) to be added."


What about in the case there's a cancellation fee configured, but refunds are not turned on?  
In that case, let's say there's a $15 cancellation fee and registration is $100
They register & pay $100
Then they cancel
A $15 fee is added to the invoice as a line item, added to the total, and added to the balance
(refunds are not turned on)
Their invoice correctly states that they owe $15... and the total is $115 which is the full amount paid (once they pay the cancellation fee).


In the scenario that manual refunds are turned on:
They register & pay $100
Then they cancel
A $15 fee is added to the invoice as a line item, added to the total, and added to the balance
The invoice correctly shows that they owe $15
An admin refunds them $100 (which really results in a refund of $85... but they "paid" $15 in a sense)


In the scenario that manual refunds are turned on:
They register & pay $100
Then they cancel
A $15 fee is added to the invoice as a line item, added to the total, and added to the balance
They are automatically refunded $85... but they "paid" $15 in a sense... and ultimately they did pay $15


Do we want to remove it from the total when we do a refund that deducts the cancellation fee? That way, if refunds are turned off (or if manual refunds are turned on & admin chooses not to refund) the invoice correctly reflects the fact they owe the fee, but is removed from the invoice in the case it's deducted from the invoice since they never really payed it... they just had a reduced refund?
Maybe we should talk about what needs to happen in our call tomorrow?

shandy-bixly commented:
> Ideally, when a user cancels a registration, the total ($20) should remain unchanged instead of having the cancellation fee ($2) to be added.
> 
> ![image](https://user-images.githubusercontent.com/106448/246883530-87f61425-014f-45a1-ac48-32410c5336d1.png)
> 
> if refund is allowed, a refund object would be created, with the amount to refund ($18) and cancellation fee ($2). In the case of the manual refund, an admin can go to the invoice view, and process the refund. The Refund Amount and Cancellation Fee would be auto-populated and editable to admin. The sum of the Refund Amount and Cancellation Fee should not be greater than the amount paid. Also a Notes field would be nice.
> 
> Example:
> 
> ![image](https://user-images.githubusercontent.com/106448/246888840-f11959af-85ce-47a8-b546-48d7991599a8.png)
> 
> @shandy-bixly @ndavis2, thoughts?

I'm not opposed to changing it if we do want to give admins the choice to not deduct cancellation fees. If we do, then I will also need to modify this wording: `To be deducted when refund is larger than pending cancellation fees.` from the bottom of the form as it won't really be true anymore

jennyq commented:
> What about in the case there's a cancellation fee configured, but refunds are not turned on?

I think the invoice table needs to add a "cancellation_fee" field. Then the cancellation fee can be stored in there and make it editable on payment refund and adjustment for admin. Apparently, the refund and adjustment forms need to be validated so that the sum of refund or adjusted amount plus the cancellation fee should not be larger than the total amount. 

It just doesn't feel right when your $100 invoice suddenly becomes $115 after you canceled your registration.  

Yes, let's discuss it in our call tomorrow.

shandy-bixly commented:
Another question. I'm adding Total Refunded to the invoice overview report. Do we want to have a pie chart for refunds as well? If so, what colors do we want to use? Maybe we can talk about that tomorrow as well?

shandy-bixly commented:
> Another question. I'm adding Total Refunded to the invoice overview report. Do we want to have a pie chart for refunds as well? If so, what colors do we want to use? Maybe we can talk about that tomorrow as well?

Or, I see now you had a comment to subtract the total refunded. In that case, we wouldn't need to add a new pie chart.

jennyq commented:
Hi Shandy,

The setting "Allow refunds" is set to Yes. 
After the registration was cancelled by the user, admin went in to refund. The Cancellation fees field is missing from the form.

![image](https://github.com/tendenci/tendenci/assets/106448/c007116a-bcb5-4a19-ac7d-7e1b233fda40)

jennyq commented:
@ndavis2,  @shandy-bixly,

Should "Payment/Credits" be adjusted after a refund happened? So, in this case, it would be $20 - $18 = $2, instead of $20.

![image](https://github.com/tendenci/tendenci/assets/106448/fad504c1-b123-47ce-a8f5-48ffb2e69507)

jennyq commented:
Hi Shandy,

Sorry but here is another hiccup - The Cancellation fee is not deducted when cancelling one of the registrants (registered two).  (Auto refund is on)
![image](https://github.com/tendenci/tendenci/assets/106448/6c4915ee-711d-44a7-8756-af71e8d39a63)

shandy-bixly commented:
> Hi Shandy,
> 
> Sorry but here is another hiccup - The Cancellation fee is not deducted when cancelling one of the registrants (registered two). (Auto refund is on) ![image](https://user-images.githubusercontent.com/106448/248412965-6c4915ee-711d-44a7-8756-af71e8d39a63.png)

I didn't think about this case. I wrote it so that it applied the cancellation fees at the very end.. but to cover this case I will need to add a field to track how much of the cancellation fees have already been applied. This shouldn't be a big change

shandy-bixly commented:
> @ndavis2, @shandy-bixly,
> 
> Should "Payment/Credits" be adjusted after a refund happened? So, in this case, it would be $20 - $18 = $2, instead of $20.
> 
> ![image](https://user-images.githubusercontent.com/106448/248410329-fad504c1-b123-47ce-a8f5-48ffb2e69507.png)

My thoughts were that they did make a $20 payment.... and so instead of deducting refunds, I added Refunds to the totals... but maybe we should talk to Nicole about which way would make the most sense?

shandy-bixly commented:
> Hi Shandy,
> 
> The setting "Allow refunds" is set to Yes. After the registration was cancelled by the user, admin went in to refund. The Cancellation fees field is missing from the form.
> 
> ![image](https://user-images.githubusercontent.com/106448/248407597-c007116a-bcb5-4a19-ac7d-7e1b233fda40.png)

So this is one we probably need to talk about to. At this point the user would have already seen a message saying that a $2 cancellation fee would apply ( since they have already cancelled). Do we want them to adjust the cancellation fee after the user has already seen the message (ie if they've already canceled) or in all cases?

jennyq commented:
> > @ndavis2, @shandy-bixly,
> > Should "Payment/Credits" be adjusted after a refund happened? So, in this case, it would be $20 - $18 = $2, instead of $20.
> > ![image](https://user-images.githubusercontent.com/106448/248410329-fad504c1-b123-47ce-a8f5-48ffb2e69507.png)
> 
> My thoughts were that they did make a $20 payment.... and so instead of deducting refunds, I added Refunds to the totals... but maybe we should talk to Nicole about which way would make the most sense?

Nicole, any thoughts on that? When a payment is voided, the `payments_credits` does get reduced https://github.com/tendenci/tendenci/blob/master/tendenci/apps/invoices/models.py#L415. So, I think same logic should apply here for refund.

jennyq commented:
> > Hi Shandy,
> > The setting "Allow refunds" is set to Yes. After the registration was cancelled by the user, admin went in to refund. The Cancellation fees field is missing from the form.
> > ![image](https://user-images.githubusercontent.com/106448/248407597-c007116a-bcb5-4a19-ac7d-7e1b233fda40.png)
> 
> So this is one we probably need to talk about to. At this point the user would have already seen a message saying that a $2 cancellation fee would apply ( since they have already cancelled). Do we want them to adjust the cancellation fee after the user has already seen the message (ie if they've already canceled) or in all cases?

Seems like this would involve too much of additional work. If Nicole agrees, I'll say let's move on.

shandy-bixly commented:
@jennyq I had to go back & modify the migration that adds adjusted_cancellation_fees to allow blank & None. This is so that if in the Refund menu an admin user can edit the cancellation fees to be $0 if needed. Just mentioning it because you might have issues with migrations since you ran the original one... Adding a new one with AlterField doesn't work because, if the default is 0 on older invoices, that could be breaking on those invoices.

jennyq commented:
> @jennyq I had to go back & modify the migration that adds adjusted_cancellation_fees to allow blank & None. This is so that if in the Refund menu an admin user can edit the cancellation fees to be $0 if needed. Just mentioning it because you might have issues with migrations since you ran the original one... Adding a new one with AlterField doesn't work because, if the default is 0 on older invoices, that could be breaking on those invoices.

Thank you Shandy for the heads up!

jennyq commented:
@shandy-bixly,

In this case, a user registered along with 2 guests. One of the guests is canceled and refunded (Cancellation Fees $2, Refunded $18). 

![image](https://github.com/tendenci/tendenci/assets/106448/6d694ed4-553f-4c17-9d4a-a86a1d6a751f)

When an admin went to refund and checked "Cancel Registration?", the "Refund Amount" was automatically set to 38 and "Cancellation Fees" 4. Shouldn't they supposed to be 36 and 6, respectively?

![image](https://github.com/tendenci/tendenci/assets/106448/c057d5fb-267e-48a3-a548-a044688333e2)

shandy-bixly commented:
> @shandy-bixly,
> 
> In this case, a user registered along with 2 guests. One of the guests is canceled and refunded (Cancellation Fees $2, Refunded $18).
> 
> ![image](https://user-images.githubusercontent.com/106448/250327439-6d694ed4-553f-4c17-9d4a-a86a1d6a751f.png)
> 
> When an admin went to refund and checked "Cancel Registration?", the "Refund Amount" was automatically set to 38 and "Cancellation Fees" 4. Shouldn't they supposed to be 36 and 6, respectively?
> 
> ![image](https://user-images.githubusercontent.com/106448/250327815-c057d5fb-267e-48a3-a548-a044688333e2.png)

so here's what's going on. Since the other 2 aren't cancelled yet, it's trying to guess what the potential cancellation fees should be...it is guessing based on 1 cancellation fee... so it's taking current cancellation fees applied + 1 new cancellation fee (since we're cancelling something). 
 I think the fix here is to update it to make it's "guess" based on the number of registrants remaining... so I need to pass in how many registrants remain so that it knows there's 2 potential cancellations & not just 1... that should be a small fix actually.

shandy-bixly commented:
> > @shandy-bixly,
> > In this case, a user registered along with 2 guests. One of the guests is canceled and refunded (Cancellation Fees $2, Refunded $18).
> > ![image](https://user-images.githubusercontent.com/106448/250327439-6d694ed4-553f-4c17-9d4a-a86a1d6a751f.png)
> > When an admin went to refund and checked "Cancel Registration?", the "Refund Amount" was automatically set to 38 and "Cancellation Fees" 4. Shouldn't they supposed to be 36 and 6, respectively?
> > ![image](https://user-images.githubusercontent.com/106448/250327815-c057d5fb-267e-48a3-a548-a044688333e2.png)
> 
> @jennyq so here's what's going on. Since the other 2 aren't cancelled yet, it's trying to guess what the potential cancellation fees should be...it is guessing based on 1 additional cancellation fee... so it's taking current cancellation fees applied + 1 new cancellation fee (since we're cancelling something). I think the fix here is to update it to make its "guess" based on the number of registrants remaining... so I need to pass in how many registrants remain so that it knows there's 2 potential cancellations & not just 1... that should be a small fix actually.

@jennyq Pushed up that fix!

jennyq commented:
Everything looks good now. Great job Shandy!

When time permits, please consider to add docs and tests for this feature. Thank you!

shandy-bixly commented:
> Everything looks good now. Great job Shandy!
> 
> When time permits, please consider to add docs and tests for this feature. Thank you!

ok will do!
